bson/src/main/org/bson/codecs/NumberCodecHelper.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.BsonInvalidOperationException;
+import org.bson.BsonReader;
+import org.bson.BsonType;
+
+import static java.lang.String.format;
+
+final class NumberCodecHelper {
+
+    static final double DEFAULT_DELTA = 0.00000000000001d;
+
+    @SuppressWarnings("unchecked")
+    static <T extends Number> T decodeNumber(final BsonReader reader, final Class<T> clazz, final double delta) {
+        Number number;
+        boolean isADownConversion;
+        BsonType bsonType = reader.getCurrentBsonType();
+        switch (bsonType) {
+            case DOUBLE:
+                number = reader.readDouble();
+                isADownConversion = clazz != Double.class;
+                break;
+            case INT64:
+                number = reader.readInt64();
+                isADownConversion = clazz == Integer.class;
```

Comment:
Conversion to double can lose information also. There are certain very large INT64 numbers that will lose precision when converted to double.